go-controller/pkg/cni/helper_linux.go

```diff
@@ -728,18 +728,39 @@ func (*defaultPodRequestInterfaceOps) UnconfigureInterface(pr *PodRequest, ifInf
 	if !ifInfo.IsDPUHostMode {
 		var err error
 		// host side interface deletion
-		hostIfName := pr.SandboxID[:(15-len(ifnameSuffix))] + ifnameSuffix
+		var hostIfName string
+		if !util.IsNetworkSegmentationSupportEnabled() || isSecondary {
+			// this is a secondary network (not primary) or segmentation is not enabled
+			hostIfName = pr.SandboxID[:(15-len(ifnameSuffix))] + ifnameSuffix
+		}
 		if pr.CNIConf.DeviceID != "" {
 			hostIfName, err = util.GetFunctionRepresentorName(pr.CNIConf.DeviceID)
 			if err != nil {
 				klog.Errorf("Failed to get the representor name for DeviceID %s for pod %s: %v",
 					pr.CNIConf.DeviceID, podDesc, err)
 			}
 		}
+		portList, err := ovsFind("interface", "name", "external-ids:sandbox="+pr.SandboxID)
+		if err != nil {
+			return fmt.Errorf("failed to list interfaces in OVS during delete for sandbox: %s, err: %w",
+				pr.SandboxID, err)
+		}
+		// hostIfName is not empty if using device ID, a secondary network, or segmentation not enabled
+		// delete the port in traditional fashion
 		if hostIfName != "" {
-			pr.deletePorts(hostIfName, pr.PodNamespace, pr.PodName)
+			pr.deletePort(hostIfName, pr.PodNamespace, pr.PodName)
+		} else {
+			// this is a primary interface deletion and segmentation is enabled, delete all ports
+			// delete happens in reverse order for attached networks, so this is the final deletion
```

Comment:
I don't follow the ordering logic mentioned here.
If I have a UDN pod, I get a delete pod:
are you saying its guaranteed ordering triggered for CNIAdd is going to be: (a) delete default interface (b) delete primary (c) delete secondary interface and thus CNIDel will be reverse so multus would have removed the secondary interfaces leaving behind only udn and default interfaces here?
I thought you are simply nuking all 3 interfaces together using deletePorts?

can you please add a test for this where we have a pod with bunch of interfaces and ordering is respected?
also a test for delete nad before pod case

Reply:
for what it's worth, from a Multus + net-attach-def standpoint, it happens in reverse: https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/pkg/multus/multus.go#L448-L455

so from a multus standpoint this is dependent on the network selection element, e.g. what is the order of `k8s.v1.cni.cncf.io/networks` and the items in that list